### PR TITLE
Initial additional information about the invariant macros and hierarchy.

### PR DESCRIPTION
**Pull request by sonodtt**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Initial additional information about the invariant macros and hierarchy.

### CODE REVIEW

**Review by chrisr-diffblue — APPROVED** [review]

**Review by xbauch — APPROVED** [review]

File: src/util/invariant.h [hunk 1]
```diff
@@ -82,6 +82,9 @@ Author: Martin Brain, martin.brain@diffblue.com
 /// family of macros, allowing constructs like
 /// `INVARIANT(x==y, my_invariantt, (T1)actual1, (T2)actual2, ...)`
 ///
+/// \ref invariant_failedt is also the base class of any 'structured
+/// exceptions' - as found in file base_exceptions.h
```
Comment (peterschrammel) [inline src/util/invariant.h]: \ref base_exceptions.h

**Review by peterschrammel — APPROVED** [review]

File: src/util/invariant.h [hunk 2]
```diff
@@ -196,6 +199,9 @@ std::string get_backtrace();
 
 void report_exception_to_stderr(const invariant_failedt &);
 
+/// This class is the backbone of all the invariant types. Every instance of an
+/// invariant is ultimately generated by this class, which is at times called
+/// via a wrapper function.
```
Comment (tautschnig) [inline src/util/invariant.h]: I am not sure this comment is in the right place? It's now part of the documentation of a method, not a class?

File: src/util/invariant.h [hunk 3]
```diff
@@ -230,8 +236,9 @@ CBMC_NORETURN
   abort();
 }
 
-/// Takes a backtrace, constructs an invariant_violatedt from reason and the
-/// backtrace, aborts printing the invariant's description.
+/// This is a wrapper class used by the macro 'INVARIANT(CONDITION, REASON)'
```
Comment (tautschnig) [inline src/util/invariant.h]: Again, I am wondering whether this is in the right place - it's documenting a method, but talks about a class?

File: src/util/invariant.h [hunk 4]
```diff
@@ -342,6 +349,8 @@ std::string assemble_diagnostics(Diagnostics &&... args)
 }
 } // namespace detail
 
+
+/// This is a wrapper class, used by the macro 'INVARIANT_WITH_DIAGNOSTICS'
```
Comment (tautschnig) [inline src/util/invariant.h]: Should it just be "wrapper function"?

File: src/util/invariant.h (hunk 2 above)
Comment (sonodtt) [reply src/util/invariant.h]: Yes - whoops! Function.

File: src/util/invariant.h (hunk 3 above)
Comment (sonodtt) [reply src/util/invariant.h]: Argh! Fixed.

File: src/util/invariant.h (hunk 4 above)
Comment (sonodtt) [reply src/util/invariant.h]: Yes. Function.

**Comment by peterschrammel** [comment]
@sonodtt, can you fix the overlong lines please: https://travis-ci.org/diffblue/cbmc/jobs/455518975#L493

**Comment by peterschrammel** [comment]
@sonodtt, clang-format is still complaining: https://travis-ci.org/diffblue/cbmc/jobs/455856034#L536

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 6717df39bf957cd82257d9583a40e3fe4ce4547b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91681970

**Pull request merged**